src/mds/Server.cc

```diff
@@ -2531,7 +2531,30 @@ void Server::dispatch_client_request(MDRequestRef& mdr)
     respond_to_request(mdr, mdr->more()->peer_error);
     return;
   }
-  
+
+  Session *session = mds->get_session(req);
+  if (session && !session->info.has_feature(CEPHFS_FEATURE_ALTERNATE_NAME) &&
+      req->get_op() != CEPH_MDS_OP_GETATTR &&
+      req->get_op() != CEPH_MDS_OP_READDIR &&
+      req->get_op() != CEPH_MDS_OP_UNLINK &&
+      req->get_op() != CEPH_MDS_OP_RMDIR) {
+    CInode *cur = try_get_auth_inode(mdr, req->get_filepath().get_ino());
+    if (!cur)
+      return;
+    MutationImpl::LockOpVec lov;
+    /* We need 'As' caps for the fscrypt context */
+    lov.add_rdlock(&cur->authlock);
+    if (!mds->locker->acquire_locks(mdr, lov)) {
+      return;
+    }
+    if (!cur->get_inode()->fscrypt_auth.empty()) {
+      dout(10) << "blocking '" << ceph_mds_op_name(req->get_op())
+	       << "' operation in encrypted node" << dendl;
+      respond_to_request(mdr, -CEPHFS_EROFS);
+      return;
+    }
```

Comment:
Should we release the `cur->authlock` here ?

Reply:
I don't really know the answer to that question :-(

I've looked into similar patterns in the code and I assumed that the `respond_to_request` would do the right thing.  But I could be wrong.  What would you suggest?  Simply doing `mds->locker->drop_locks(mdr.get())` before the `respond_to_request()` call?

Reply:
Yeah, the `respond_to_request()` will help release it.

But in case the fscrypt is not enabled then the code will continue to run the following code with the `authlock` locked, what will happen if it tries to acquire this lock again later ?

Reply:
Ok, makes sense.  Thanks, I'll add this lock release in the next revision.